### PR TITLE
[CI] Add clang-tidy to lint

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
`clang-tidy` has checks to detect many bug prone patterns in C++. It can also enforce style rules in addition to `clang-format`.

A set of checks are enabled, referencing `grpc/grpc` and a few other projects. I tested these checks on a few files to disable the most frequently triggered ones that are relatively benign e.g. `misc-unused-parameters`. We can do a cleanup and re-enable some of these checks later.

The next steps would be to enable additional useful checks e.g. for readability, and disable noisy ones. Another issue is because of the need to rebuild with `clang`, `lint` takes 20 min longer to run which I hope to fix later.

### Using clang-tidy on CI
CI runs `clang-tidy` automatically as part of the `lint` process. If there is any issue, the `lint` process would produce an error. PR author just need to check the `lint` output and fix the issues.

### Using clang-tidy locally for a PR
With `clang` and `clang-tidy` (version 12) installed and available on `PATH`, PR authors can run `scripts/check-git-clang-tidy-output.sh` at repo root to check for `clang-tidy` issue in the current PR.

### Using clang-tidy locally to fix a pattern across the project
- Ensure `clang` and `clang-tidy` (version 12) are available on `PATH`. Run `scripts/check-git-clang-tidy-output.sh` at repo root to generate `compile_commands.json`.
- Download [run-clang-tidy.py](https://github.com/llvm-mirror/clang-tools-extra/blob/master/clang-tidy/tool/run-clang-tidy.py) to repo root.
- Run the script to detect and generate fixes for one or more checks. e.g. `./run-clang-tidy.py -fix -checks=-*,modernize-make-unique`

## Related issue number
#388

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simon-mo — APPROVED** [review]

**Review by scv119 — APPROVED** [review]
image

File: .clang-tidy [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Disabled checks are due to frequent false positives and noisiness.
+# They can be re-enabled later on if the benefit outweighs the cost.
+# TODO: enable google-* and readability-* families of checks.
+Checks: >
+  abseil-*,
```
[the remark this answers is not in the document]
Comment (mwtian) [reply .clang-tidy:5]: All of the explicitly disabled checks are kind of too noisy right now, e.g. during a refactor. For a followup PR, I only have plans to add selected additional checks from google-*, performance-*, cppcoreguidelines-* and readability-*.

**Comment by rkooo567** [comment]
Is it possible to make `scripts/check-git-clang-tidy-output.sh` run when we run `./ci/travis.format.sh`?

File: .clang-tidy [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  -misc-unused-parameters,
+  modernize-*,
+  -modernize-avoid-c-arrays,
+  -modernize-pass-by-value,
+  -modernize-return-braced-init-list,
+  -modernize-use-emplace,
+  -modernize-use-trailing-return-type,
+
+# Turn all the warnings from the checks above into errors.
+WarningsAsErrors: "*"
```
Comment (rkooo567) [inline .clang-tidy:19]: New line

File: ci/generate_compile_commands/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+)
+
+extra_action(
+    name = "compile_command_action",
+    cmd = "$(location :extract_compile_command) \
+        $(EXTRA_ACTION_FILE) \
+        $(output $(ACTION_ID).compile_command.json)",
+    out_templates = ["$(ACTION_ID).compile_command.json"],
+    tools = [":extract_compile_command"],
+)
```
Comment (rkooo567) [inline ci/generate_compile_commands/BUILD.bazel:32]: new line on new files

**Review by rkooo567 — APPROVED** [review]
nit comments. I think the only requirement is https://github.com/ray-project/ray/pull/18124#issuecomment-908104387. I assume it shouldn't be too hard to add it?

**Comment by mwtian** [comment]
> nit comments. I think the only requirement is [#18124 (comment)](https://github.com/ray-project/ray/pull/18124#issuecomment-908104387). I assume it shouldn't be too hard to add it?

Yeah, I think there is a bit more work and considerations needed. Currently running `clang-tidy` requires `clang` to be installed and takes 10+ minutes to build the Ray source code. I think the step is too expensive to be added to `scripts/format.h`. I want to try one or both of the following first:
(1) Use a faster approach to generate the compile commands used by `clang-tidy`.
(2) Switch to compile with `clang` by default, so `clang-tidy` does not need recompiling.

Then we can run `clang-tidy` in the format script or pre-merge hook.

**Comment by mwtian** [comment]
Timeout in Windows looks unrelated.

**Comment by rkooo567** [comment]
TODO is fine, but the instruction `With clang and clang-tidy (version 12) installed and available on PATH, PR authors can run scripts/check-git-clang-tidy-output.sh at repo root to check for clang-tidy issue in the current PR.` is not easily discoverable  and can be a blocker for new contributors or us (who don't know the existence of this). 

Can you 
- Update the contributor doc to run this script when clang tidy is failed on CI
- Modify a script a little bit to print to run `check-git-clang-tidy-output.sh` when it fails?

**Comment by rkooo567** [comment]
Btw, changing the compiler to clang sounds great. This can also probably helps us building asan wheels (I saw gcc static linking asan implementation is fragile whereas that's the default option for clang cc @scv119 )

**Comment by rkooo567** [comment]
@mwtian would you have some time to address the last comment, or should we handle in the next PR? (I prefer they are done within this PR, but lmk what you think! I don't want to block PRs too long)

**Comment by mwtian** [comment]
> @mwtian would you have some time to address the last comment, or should we handle in the next PR? (I prefer they are done within this PR, but lmk what you think! I don't want to block PRs too long)

Yes, as we discussed offline last time, I would make the clang change in a separate PR, and then come back to this!

**Comment by mwtian** [comment]
Added information for contributors, enhanced the error message in `ci/travis/check-git-clang-tidy-output.sh` and added notice for different OS that the script uses `clang` as compiler.

**Comment by mwtian** [comment]
Windows failure seems unrelated.

**Pull request merged**